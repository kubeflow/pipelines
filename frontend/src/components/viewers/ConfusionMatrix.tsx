/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Viewer, { ViewerConfig, PlotType } from './Viewer';
import { color, commonCss, fontsize } from '../../Css';
import { classes, stylesheet } from 'typestyle';

const legendNotches = 5;

export interface ConfusionMatrixConfig extends ViewerConfig {
  data: number[][];
  axes: string[];
  labels: string[];
  type: PlotType;
}

interface ConfusionMatrixProps {
  configs: ConfusionMatrixConfig[];
  maxDimension?: number;
}

interface ConfusionMatrixState {
  activeCell: [number, number];
}

class ConfusionMatrix extends Viewer<ConfusionMatrixProps, ConfusionMatrixState> {
  private _opacities: number[][] = [];
  private _config = this.props.configs[0];
  private _max =
    this._config &&
    Math.max(...this._config.data.map(d => d.map(n => +n)).map(d => Math.max(...d)));
  private _minRegularCellDimension = 15;
  private _maxRegularCellDimension = 80;
  private _cellDimension = this._config
    ? Math.max(
        Math.min(
          (this.props.maxDimension || 700) / this._config.data.length,
          this._maxRegularCellDimension,
        ),
        this._minRegularCellDimension,
      ) - 1
    : 0;
  private _shrinkThreshold = 600;

  private _css = stylesheet({
    activeLabel: {
      borderRadius: 5,
      color: color.theme,
      fontWeight: 'bold',
    },
    cell: {
      border: 'solid 1px ' + color.background,
      fontSize: this._isSmall() ? fontsize.small : fontsize.base,
      height: this._cellDimension,
      minHeight: this._cellDimension,
      minWidth: this._cellDimension,
      position: 'relative',
      textAlign: 'center',
      verticalAlign: 'middle',
      width: this._cellDimension,
    },
    legend: {
      background: `linear-gradient(${color.theme}, ${color.background})`,
      borderRight: 'solid 1px #777',
      marginLeft: 20,
      minWidth: 10,
      position: 'relative',
      width: 10,
    },
    legendLabel: {
      left: 15,
      position: 'absolute',
      top: -7,
    },
    legendNotch: {
      borderTop: 'solid 1px #777',
      left: '100%',
      paddingLeft: 5,
      position: 'absolute',
      width: 5,
    },
    overlay: {
      backgroundColor: '#000',
      bottom: 0,
      left: 0,
      opacity: 0,
      position: 'absolute',
      right: 0,
      top: 0,
    },
    root: {
      flexGrow: 1,
      justifyContent: 'center',
      margin: 'auto',
      pointerEvents: this._isSmall() ? 'none' : 'initial', // Disable interaction for snapshot view
      position: 'relative',
      width: 'fit-content',
    },
    xAxisLabel: {
      color: color.foreground,
      fontSize: 15,
      fontWeight: 'bold',
      paddingLeft: 20,
      position: 'absolute',
    },
    xlabel: {
      marginLeft: 15,
      overflow: 'hidden',
      position: 'absolute',
      textAlign: 'left',
      textOverflow: 'ellipsis',
      transform: 'rotate(60deg)',
      transformOrigin: 'left',
      whiteSpace: 'nowrap',
      width: 150,
    },
    yAxisLabel: {
      color: color.foreground,
      fontSize: 15,
      height: 25,
      paddingRight: 20,
      textAlign: 'right',
    },
    ylabel: {
      lineHeight: `${this._cellDimension}px`,
      marginRight: 10,
      minWidth: this._cellDimension,
      textAlign: 'right',
      whiteSpace: 'nowrap',
    },
  });

  constructor(props: any) {
    super(props);

    if (!this._config) {
      return;
    }

    this.state = {
      activeCell: [-1, -1],
    };

    for (const i of this._config.data) {
      const row = [];
      for (const j of i) {
        row.push(+j / this._max);
      }
      this._opacities.push(row);
    }
  }

  public getDisplayName(): string {
    return 'Confusion matrix';
  }

  public render(): JSX.Element | null {
    if (!this._config) {
      return null;
    }

    const [activeRow, activeCol] = this.state.activeCell;
    const [xAxisLabel, yAxisLabel] = this._config.axes;
    const small = this._isSmall();

    return (
      <div className={classes(commonCss.flex, this._css.root)}>
        <table>
          <tbody>
            {!small && (
              <tr>
                <td className={this._css.yAxisLabel}>{yAxisLabel}</td>
              </tr>
            )}
            {this._config.data.map((row, r) => (
              <tr key={r}>
                {!small && (
                  <td>
                    <div
                      className={classes(
                        this._css.ylabel,
                        r === activeRow ? this._css.activeLabel : '',
                      )}
                    >
                      {this._config.labels[r]}
                    </div>
                  </td>
                )}
                {row.map((cell, c) => (
                  <td
                    key={c}
                    className={this._css.cell}
                    style={{
                      backgroundColor: `rgba(41, 121, 255, ${this._opacities[r][c]})`,
                      color: this._opacities[r][c] < 0.6 ? color.foreground : color.background,
                    }}
                    onMouseOver={() => this.setState({ activeCell: [r, c] })}
                  >
                    <div
                      className={this._css.overlay}
                      style={{
                        opacity: r === activeRow || c === activeCol ? 0.05 : 0,
                      }}
                    />
                    {cell}
                  </td>
                ))}
              </tr>
            ))}

            {/* Footer */}
            {!small && (
              <tr>
                <th className={this._css.xlabel} />
                {this._config.labels.map((label, i) => (
                  <th key={i}>
                    <div
                      className={classes(
                        i === activeCol ? this._css.activeLabel : '',
                        this._css.xlabel,
                      )}
                    >
                      {label}
                    </div>
                  </th>
                ))}
                <td className={this._css.xAxisLabel}>{xAxisLabel}</td>
              </tr>
            )}
          </tbody>
        </table>

        {!small && (
          <div
            className={this._css.legend}
            style={{ height: 0.75 * this._config.data.length * this._cellDimension }}
          >
            <div className={this._css.legendNotch} style={{ top: 0 }}>
              <span className={this._css.legendLabel}>{this._max}</span>
            </div>
            {new Array(legendNotches).fill(0).map((_, i) => (
              <div
                key={i}
                className={this._css.legendNotch}
                style={{ top: ((legendNotches - i) / legendNotches) * 100 + '%' }}
              >
                <span className={this._css.legendLabel}>
                  {Math.floor((i / legendNotches) * this._max)}
                </span>
              </div>
            ))}
          </div>
        )}
      </div>
    );
  }

  private _isSmall(): boolean {
    return !!this.props.maxDimension && this.props.maxDimension < this._shrinkThreshold;
  }
}

export default ConfusionMatrix;
