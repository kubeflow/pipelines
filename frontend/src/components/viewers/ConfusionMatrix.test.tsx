/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { shallow } from 'enzyme';
import ConfusionMatrix, { ConfusionMatrixConfig } from './ConfusionMatrix';
import { PlotType } from './Viewer';

describe('ConfusionMatrix', () => {
  it('does not break on empty data', () => {
    const tree = shallow(<ConfusionMatrix configs={[]} />);
    expect(tree).toMatchSnapshot();
  });

  const data = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
  ];
  const config: ConfusionMatrixConfig = {
    axes: ['test x axis', 'test y axis'],
    data,
    labels: ['label1', 'label2'],
    type: PlotType.CONFUSION_MATRIX,
  };
  it('renders a basic confusion matrix', () => {
    const tree = shallow(<ConfusionMatrix configs={[config]} />);
    expect(tree).toMatchSnapshot();
  });

  it('does not break on asymetric data', () => {
    const testConfig = { ...config };
    testConfig.data = data.slice(1);
    const tree = shallow(<ConfusionMatrix configs={[testConfig]} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders only one of the given list of configs', () => {
    const tree = shallow(<ConfusionMatrix configs={[config, config, config]} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders a small confusion matrix snapshot, with no labels or footer', () => {
    const tree = shallow(<ConfusionMatrix configs={[config]} maxDimension={100} />);
    expect(tree).toMatchSnapshot();
  });

  it('activates row/column on cell hover', () => {
    const tree = shallow(<ConfusionMatrix configs={[config]} />);
    tree
      .find('td')
      .at(2)
      .simulate('mouseOver');
    expect(tree.state()).toHaveProperty('activeCell', [0, 0]);
  });

  it('returns a user friendly display name', () => {
    expect(ConfusionMatrix.prototype.getDisplayName()).toBe('Confusion matrix');
  });
});
