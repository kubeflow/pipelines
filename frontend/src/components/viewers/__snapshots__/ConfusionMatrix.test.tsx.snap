// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfusionMatrix does not break on asymetric data 1`] = `
<div
  className="flex root"
>
  <table>
    <tbody>
      <tr>
        <td
          className="yAxisLabel"
        >
          test y axis
        </td>
      </tr>
      <tr
        key="0"
      >
        <td>
          <div
            className="ylabel"
          >
            label1
          </div>
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.375)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          3
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.5)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          4
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.625)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          5
        </td>
      </tr>
      <tr
        key="1"
      >
        <td>
          <div
            className="ylabel"
          >
            label2
          </div>
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.75)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          6
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.875)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          7
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 1)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          8
        </td>
      </tr>
      <tr>
        <th
          className="xlabel"
        />
        <th
          key="0"
        >
          <div
            className="xlabel"
          >
            label1
          </div>
        </th>
        <th
          key="1"
        >
          <div
            className="xlabel"
          >
            label2
          </div>
        </th>
        <td
          className="xAxisLabel"
        >
          test x axis
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="legend"
    style={
      Object {
        "height": 118.5,
      }
    }
  >
    <div
      className="legendNotch"
      style={
        Object {
          "top": 0,
        }
      }
    >
      <span
        className="legendLabel"
      >
        8
      </span>
    </div>
    <div
      className="legendNotch"
      key="0"
      style={
        Object {
          "top": "100%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        0
      </span>
    </div>
    <div
      className="legendNotch"
      key="1"
      style={
        Object {
          "top": "80%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        1
      </span>
    </div>
    <div
      className="legendNotch"
      key="2"
      style={
        Object {
          "top": "60%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        3
      </span>
    </div>
    <div
      className="legendNotch"
      key="3"
      style={
        Object {
          "top": "40%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        4
      </span>
    </div>
    <div
      className="legendNotch"
      key="4"
      style={
        Object {
          "top": "20%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        6
      </span>
    </div>
  </div>
</div>
`;

exports[`ConfusionMatrix does not break on empty data 1`] = `""`;

exports[`ConfusionMatrix renders a basic confusion matrix 1`] = `
<div
  className="flex root"
>
  <table>
    <tbody>
      <tr>
        <td
          className="yAxisLabel"
        >
          test y axis
        </td>
      </tr>
      <tr
        key="0"
      >
        <td>
          <div
            className="ylabel"
          >
            label1
          </div>
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          0
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.125)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          1
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.25)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          2
        </td>
      </tr>
      <tr
        key="1"
      >
        <td>
          <div
            className="ylabel"
          >
            label2
          </div>
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.375)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          3
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.5)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          4
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.625)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          5
        </td>
      </tr>
      <tr
        key="2"
      >
        <td>
          <div
            className="ylabel"
          />
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.75)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          6
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.875)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          7
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 1)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          8
        </td>
      </tr>
      <tr>
        <th
          className="xlabel"
        />
        <th
          key="0"
        >
          <div
            className="xlabel"
          >
            label1
          </div>
        </th>
        <th
          key="1"
        >
          <div
            className="xlabel"
          >
            label2
          </div>
        </th>
        <td
          className="xAxisLabel"
        >
          test x axis
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="legend"
    style={
      Object {
        "height": 177.75,
      }
    }
  >
    <div
      className="legendNotch"
      style={
        Object {
          "top": 0,
        }
      }
    >
      <span
        className="legendLabel"
      >
        8
      </span>
    </div>
    <div
      className="legendNotch"
      key="0"
      style={
        Object {
          "top": "100%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        0
      </span>
    </div>
    <div
      className="legendNotch"
      key="1"
      style={
        Object {
          "top": "80%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        1
      </span>
    </div>
    <div
      className="legendNotch"
      key="2"
      style={
        Object {
          "top": "60%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        3
      </span>
    </div>
    <div
      className="legendNotch"
      key="3"
      style={
        Object {
          "top": "40%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        4
      </span>
    </div>
    <div
      className="legendNotch"
      key="4"
      style={
        Object {
          "top": "20%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        6
      </span>
    </div>
  </div>
</div>
`;

exports[`ConfusionMatrix renders a small confusion matrix snapshot, with no labels or footer 1`] = `
<div
  className="flex root"
>
  <table>
    <tbody>
      <tr
        key="0"
      >
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          0
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.125)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          1
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.25)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          2
        </td>
      </tr>
      <tr
        key="1"
      >
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.375)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          3
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.5)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          4
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.625)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          5
        </td>
      </tr>
      <tr
        key="2"
      >
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.75)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          6
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.875)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          7
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 1)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          8
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ConfusionMatrix renders only one of the given list of configs 1`] = `
<div
  className="flex root"
>
  <table>
    <tbody>
      <tr>
        <td
          className="yAxisLabel"
        >
          test y axis
        </td>
      </tr>
      <tr
        key="0"
      >
        <td>
          <div
            className="ylabel"
          >
            label1
          </div>
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          0
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.125)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          1
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.25)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          2
        </td>
      </tr>
      <tr
        key="1"
      >
        <td>
          <div
            className="ylabel"
          >
            label2
          </div>
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.375)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          3
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.5)",
              "color": "#000",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          4
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.625)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          5
        </td>
      </tr>
      <tr
        key="2"
      >
        <td>
          <div
            className="ylabel"
          />
        </td>
        <td
          className="cell"
          key="0"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.75)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          6
        </td>
        <td
          className="cell"
          key="1"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 0.875)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          7
        </td>
        <td
          className="cell"
          key="2"
          onMouseOver={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(41, 121, 255, 1)",
              "color": "#fff",
            }
          }
        >
          <div
            className="overlay"
            style={
              Object {
                "opacity": 0,
              }
            }
          />
          8
        </td>
      </tr>
      <tr>
        <th
          className="xlabel"
        />
        <th
          key="0"
        >
          <div
            className="xlabel"
          >
            label1
          </div>
        </th>
        <th
          key="1"
        >
          <div
            className="xlabel"
          >
            label2
          </div>
        </th>
        <td
          className="xAxisLabel"
        >
          test x axis
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="legend"
    style={
      Object {
        "height": 177.75,
      }
    }
  >
    <div
      className="legendNotch"
      style={
        Object {
          "top": 0,
        }
      }
    >
      <span
        className="legendLabel"
      >
        8
      </span>
    </div>
    <div
      className="legendNotch"
      key="0"
      style={
        Object {
          "top": "100%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        0
      </span>
    </div>
    <div
      className="legendNotch"
      key="1"
      style={
        Object {
          "top": "80%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        1
      </span>
    </div>
    <div
      className="legendNotch"
      key="2"
      style={
        Object {
          "top": "60%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        3
      </span>
    </div>
    <div
      className="legendNotch"
      key="3"
      style={
        Object {
          "top": "40%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        4
      </span>
    </div>
    <div
      className="legendNotch"
      key="4"
      style={
        Object {
          "top": "20%",
        }
      }
    >
      <span
        className="legendLabel"
      >
        6
      </span>
    </div>
  </div>
</div>
`;
